File: src/main/java/medium/Q0003.java
package medium;

import java.util.HashSet;
import java.util.Set;

/**
 * 3. Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */
class Q0003 {

  // O(n) time, O(n) space, where n is the length of `s`
  // Note: Java's HashSet.contains() and HashSet.remove() are O(1). Calling Object.hashCode() on all the possible ASCII
  // characters in `s` will lead to ~uniform distribution in the underlying hash table.
  int lengthOfLongestSubstring(String s) {
    int max = 0, start = 0;
    Set<Character> visited = new HashSet<>();
    for (int end = 0; end < s.length(); end++) {
      while (visited.contains(s.charAt(end))) {
        visited.remove(s.charAt(start));
        start++;
      }
      visited.add(s.charAt(end));
      max = Math.max(max, visited.size());
    }
    return max;
  }
}

File: src/main/python/medium/Q0003.py
"""
3. Longest Substring Without Repeating Characters
https://leetcode.com/problems/longest-substring-without-repeating-characters
"""


# Time: O(n), Space: O(n), where n is the length of s
def length_of_longest_substring(s: string) -> int:
    longest = 0
    unique = set()
    l = 0
    for r in range(len(s)):
        while s[r] in unique:
            unique.remove(s[l])
            l += 1
        unique.add(s[r])
        longest = max(longest, len(unique))
    return longest
